lib/mpool/lib/mblock_file.c

```diff
@@ -896,17 +899,50 @@ mblock_file_meta_validate(const struct mblock_file *mbfp, uint64_t *mbidv, int m
     return err;
 }
 
+static merr_t
+mblock_alen_get(struct mblock_file *mbfp, uint64_t mbid, uint32_t *alen)
+{
+    off_t start, cur, end;
+    size_t mblocksz;
+
+    INVARIANT(mbfp && alen);
+
+    mblocksz = mbfp->mblocksz;
+    start = block_off(mbid, mblocksz);
+    end = start + mblocksz - 1;
+
+    cur = lseek(mbfp->fd, start, SEEK_SET);
+    assert(cur == start);
+
+    *alen = mblocksz;
+    while ((cur = lseek(mbfp->fd, cur, SEEK_HOLE)) >= 0) {
+        off_t hole_start = cur;
+
+        if (hole_start > end)
+            break;
+
+        cur = lseek(mbfp->fd, cur, SEEK_DATA);
+        if (cur > end || (cur < 0 && errno == ENXIO)) {
+            assert(hole_start <= end);
+            *alen -= (end - hole_start + 1);
+            cur = 0;
+            break;
+        }
```

Comment:
Do we need to support multiple holes in mblocks?  Am I exposing the fact that I didn't fully understand the new kvset split technique?

Reply:
This function is written in such a way that we can handle multiple holes.

For the case of split, we will have only the following possibilities:
Hole-Data
Data-Hole-Data

Reply:
What about multiple splits?

Reply:
That's a great question. I totally missed that :-)

Reply:
Suggestion: revert to heavier weight kvset split if an mblock needs to be resplit.

Reply:
E.g., ycsb native load, the initial leaf node in the system gets split 128 times!!!

Reply:
I think the algorithm described in the ticket should handle recursive splits.. Thinking through it now with some examples...

Reply:
It's probably not too common to have the same mblock be split multiple times.  In a monotonic load it would never happen.  In a uniform load it's more likely but still prob not common unless the values were large, which would result in fewer kblocks and correspondingly more resplits.

Reply:
Worked out some examples.. The algorithm described in the ticket should handle recursive vblock splits correctly. 

Here's what the algorithm does at a high level:

Let's say,
**(0, wlen - 1)**: vblock offset range
**S**: offfset in the vblock corresponding to key that follows the split-key

**left_split(vblock)**:
punch[S, wlen - footer_len) and retain the data in (0, S)

**right_split(vblock)**:
clone[S, wlen) which internally punches (0, S)

Applying the left_split(vblock) and right_split(vblock) recursively just keeps extending the holes in either direction depending on which side of the split it is.

Notes:
1. Recursive splits of the same vblock leaves the wlen intact as the vblock footer is at a fixed offset (wlen - footer_len) and cannot be touched.
2. The invariant that all keys referring a vblock is within the min, max key range stored in its footer is preserved even after this vblock undergoes multiple splits
3. The hole punch just discards the values that are not needed on either the left/right side of the split, IOW gets rid of garbage.

@gsramdasi double checked my examples. He also called out that there could only be Hole-Data-Hole-Data or a subset of it in each vblock even after infinite recursive splits as these splits just extend the existing holes in either direction.